File: Languages/cpp/algorithms-dsa/leetcode/tries/implement_trie.cpp
/**
 * LeetCode 208: Implement Trie (Prefix Tree)
 * Topic: Tries
 * Difficulty: Medium
 *
 * Implement a trie with insert, search, and startsWith methods.
 * TrieNode with children array of size 26.
 * Time: O(L) per operation where L = word length, Space: O(N*L)
 */
#include <iostream>
#include <string>
#include <cassert>
using namespace std;

struct TrieNode {
    TrieNode* children[26];
    bool isEnd;
    TrieNode() : isEnd(false) {
        for (int i = 0; i < 26; i++) children[i] = nullptr;
    }
};

class Trie {
    TrieNode* root;
public:
    Trie() { root = new TrieNode(); }

    void insert(string word) {
        TrieNode* node = root;
        for (char c : word) {
            int idx = c - 'a';
            if (!node->children[idx])
                node->children[idx] = new TrieNode();
            node = node->children[idx];
        }
        node->isEnd = true;
    }

    bool search(string word) {
        TrieNode* node = root;
        for (char c : word) {
            int idx = c - 'a';
            if (!node->children[idx]) return false;
            node = node->children[idx];
        }
        return node->isEnd;
    }

    bool startsWith(string prefix) {
        TrieNode* node = root;
        for (char c : prefix) {
            int idx = c - 'a';
            if (!node->children[idx]) return false;
            node = node->children[idx];
        }
        return true;
    }
};

int main() {
    Trie trie;
    trie.insert("apple");
    assert(trie.search("apple") == true);
    assert(trie.search("app") == false);
    assert(trie.startsWith("app") == true);
    trie.insert("app");
    assert(trie.search("app") == true);
    assert(trie.search("banana") == false);
    assert(trie.startsWith("ban") == false);

    cout << "All tests passed!" << endl;
    return 0;
}

File: Languages/go/algorithms-dsa/leetcode/tries/implement_trie.go
/*
LeetCode #208: Implement Trie (Prefix Tree)
Topic: Tries
Difficulty: Medium

Implement a trie with insert, search, and startsWith methods.

Approach: Each node has an array of 26 children (for lowercase letters)
and a boolean marking end of word. Insert builds path, search/startsWith
traverse checking existence.

Time: O(m) per operation where m is word length, Space: O(n * m)
*/
package main

import "fmt"

type TrieNode struct {
	children [26]*TrieNode
	isEnd    bool
}

type Trie struct {
	root *TrieNode
}

func ConstructorTrie() Trie {
	return Trie{root: &TrieNode{}}
}

func (t *Trie) Insert(word string) {
	node := t.root
	for _, ch := range word {
		idx := ch - 'a'
		if node.children[idx] == nil {
			node.children[idx] = &TrieNode{}
		}
		node = node.children[idx]
	}
	node.isEnd = true
}

func (t *Trie) Search(word string) bool {
	node := t.findNode(word)
	return node != nil && node.isEnd
}

func (t *Trie) StartsWith(prefix string) bool {
	return t.findNode(prefix) != nil
}

func (t *Trie) findNode(s string) *TrieNode {
	node := t.root
	for _, ch := range s {
		idx := ch - 'a'
		if node.children[idx] == nil {
			return nil
		}
		node = node.children[idx]
	}
	return node
}

func main() {
	trie := ConstructorTrie()
	trie.Insert("apple")

	if trie.Search("apple") != true {
		fmt.Println("FAIL Test 1")
	} else {
		fmt.Println("PASS Test 1")
	}

	if trie.Search("app") != false {
		fmt.Println("FAIL Test 2")
	} else {
		fmt.Println("PASS Test 2")
	}

	if trie.StartsWith("app") != true {
		fmt.Println("FAIL Test 3")
	} else {
		fmt.Println("PASS Test 3")
	}

	trie.Insert("app")
	if trie.Search("app") != true {
		fmt.Println("FAIL Test 4")
	} else {
		fmt.Println("PASS Test 4")
	}
}
